Build overcloud host images using packages from Ark

m-bull opened the pull request:
Override package mirrors for both Ubuntu Jammy and Rocky Linux 9 overcloud images so that they are built from package snapshots defined in `pulp-repo-versions.yml`.

Also build IPA from the same mirrors, although this fails currently because python3.10-dev supplied by Ark in the snapshot defined in `pulp-repo-versions.yml` is too old. Updating the package snapshot to a newer version means dependencies can be solved and the image is able to be built.

Use the authenticating Pulp proxy for everything that touches packages on Ark, to prevent credentials accidentally leaking into built artefacts.

Requires: https://github.com/stackhpc/stackhpc-image-elements/pull/67

m-bull commented on etc/kayobe/stackhpc.yml:
```diff
@@ -166,3 +166,11 @@ stackhpc_docker_registry_password: "{{ pulp_password }}"
+stackhpc_dib_image_elements_repos:
+  - repo: "https://github.com/stackhpc/stackhpc-image-elements"
+    local: "{{ source_checkout_path }}/stackhpc-image-elements"
+    version: "rocky-container-pulp-2"
```
Use this branch for testing and update when https://github.com/stackhpc/stackhpc-image-elements/pull/67 is merged and a new tag cut.

Alex-Welsh replied:
v1.6.4 tag released

m-bull commented:
TODO: Some automation to trigger a host/IPA image build when `pulp-repo-versions.yml` is updated, ?

Alex-Welsh reviewed (commented):
Very nice work. LGTM and will approve once https://github.com/stackhpc/stackhpc-image-elements/pull/67 is finalised

Alex-Welsh commented:
Thinking about it, should probably add a release note

m-bull commented:
Back to draft while I tweak this a bit.

I think it would be nicer to have switch(es) that could be set which enables this functionality, then leave all of the config in the base environment behind that switch, and just make enabling it in an environment (either CI or customer) something like:

```
stackhpc_build_overcloud_image_from_pulp_package_mirrors: true
stackhpc_build_ipa_image_from_pulp_package_mirrors: true
```

Alex-Welsh reviewed (commented):
This looks good to me, @m-bull is there anything else we have to add before we merge?

jackhodgkiss committed:
Build overcloud and IPA images from StackHPC package mirrors

jackhodgkiss committed:
Use env hooks to setup Pulp proxy in CI

Use the authenticating pulp_proxy for all CI build jobs that need packages from Ark -
host images, Kolla images and the IPA image.

MoteHue committed:
Don't build IPA from Ark

We've decided to move this to a separate PR

MoteHue committed:
Use tagged release of stackhpc-image-elements

MoteHue committed:
Use noble for DIB_DISTRIBUTION_MIRROR

MoteHue committed:
Install docker-buildx after configuring host

MoteHue committed:
Keep building from Ark behind a flag

MoteHue commented:
Tested here: https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/17768572122/job/50498257266

MoteHue commented:
> TODO: Some automation to trigger a host/IPA image build when `pulp-repo-versions.yml` is updated, ?

I plan to work on this too, but in a follow-up PR.

MoteHue committed:
Add release note

MoteHue commented:
Final testing here: https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/17769758155/job/50502404771

Alex-Welsh commented:
@MoteHue could you include newly built images with these changes in the same PR?

Alex-Welsh commented on releasenotes/notes/build-images-from-ark-6eb0d5c8779bb970.yaml:
```diff
@@ -0,0 +1,8 @@
+    This feature can be toggled with the flag
+    ``stackhpc_build_overcloud_image_from_pulp_package_mirrors``. For the CI
+    Builder environment, this is enabled, so our StackHPC overcloud host images
+    will now be built with packages from Ark.
```
Minor rewording suggestions for clarity
```suggestion
    Added support for building overcloud host images from Pulp package mirrors.
    This feature can be toggled with the flag
    ``stackhpc_build_overcloud_image_from_pulp_package_mirrors``. By default,
    this feature is enabled in the ``ci-builder`` environment and disabled
    otherwise. StackHPC overcloud host images will now be built with
    packages from Ark.
```

Alex-Welsh replied:
i'd also change `will now be built with packages from Ark` to `are now built with packages from Ark` if you include new images in the PR

MoteHue replied:
Done

Alex-Welsh commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -212,7 +212,7 @@ jobs:
           source src/kayobe-config/kayobe-env --environment ci-builder &&
           kayobe seed host command run \
-          --command "sudo apt update && sudo apt -y install gcc git libffi-dev python3-dev python-is-python3 python3-venv" --show-output
+          --command "sudo apt update && sudo apt -y install gcc git libffi-dev python3-dev python-is-python3 python3-venv containerd docker.io docker-buildx" --show-output
```
Is buildx not `docker-buildx-plugin`?

https://docs.docker.com/engine/install/ubuntu/#:~:text=sudo%20apt%2Dget%20install%20docker%2Dce%20docker%2Dce%2Dcli%20containerd.io%20docker%2Dbuildx%2Dplugin%20docker%2Dcompose%2Dplugin

MoteHue replied:
No, it would fail with `No package matching 'docker-buildx-plugin' is available` when I tried with that. It's why I install `docker-buildx` and its dependencies instead.

Alex-Welsh replied:
Interesting, but not worth investigating

MoteHue replied:
Agreed, it's just for the builder VM anyway. Doesn't propagate into our images

MoteHue committed:
New overcloud host images build from Ark packages

MoteHue commented:
> @MoteHue could you include newly built images with these changes in the same PR?

Added

MoteHue committed:
Reword release note

Alex-Welsh commented:
/gemini review

MoteHue commented on etc/kayobe/environments/ci-builder/hooks/overcloud-container-image-build/pre.d/10-pulp-auth-proxy.yml:
```diff
@@ -0,0 +1 @@
+../../../../../ansible/pulp-auth-proxy.yml
```
Gemini doesn't understand symlinks

MoteHue commented on etc/kayobe/environments/ci-builder/hooks/overcloud-host-image-build/pre.d/10-pulp-auth-proxy.yml:
```diff
@@ -0,0 +1 @@
+../../../../../ansible/pulp-auth-proxy.yml
```
Gemini doesn't understand symlinks

MoteHue commented on releasenotes/notes/build-images-from-ark-6eb0d5c8779bb970.yaml:
```diff
@@ -0,0 +1,9 @@
+    ``stackhpc_build_overcloud_image_from_pulp_package_mirrors``. By default,
+    this feature is enabled in the ``ci-builder`` environment and disabled
+    otherwise. StackHPC overcloud host images are now be built with packages
+    from Ark.
```
Fixed

Alex-Welsh commented:
Before I forget, there is a trivy scanner for VM images, but it's got pretty rubbish support at the moment (and has done for a while so I don't know if it'll ever get any better). Currently [only works on VMDK images](https://trivy.dev/latest/docs/target/vm/#virtual-machine-images), and only the [streamOptimised disk type](https://trivy.dev/latest/docs/target/vm/#vmdk-disk-types). We might be able to convert things to that, then scan them but that'd be a lot of faff. Anyways, if it gets better, we should totally scan our images as well

bbezak commented:
btw, those images were not promoted, I promoted them via GitHub actions, however the checksum of (at least) R9 image here is broken

```
curl -L https://musica:CREDS@ark.stackhpc.com/pulp/content/kayobe-images/2025.1/rocky/9/2025.1-20250916T145046/overcloud-rocky-9.qcow2.sha256
25afb927b163fd1e57d29ee8a275982e1bd4dae928c668550299859d6ac66902
```
